python/ray/utils.py

```diff
@@ -215,3 +217,32 @@ def merge_dicts(d1, d2):
     d = d1.copy()
     d.update(d2)
     return d
+
+
+class ThreadSafeProxy(object):
+
+    def __init__(self, orig_obj, lock):
+        self.orig_obj = orig_obj
+        self.lock = lock
+        self._wrapper_cache = {}
+
+    def __getattr__(self, attr):
+        orig_attr = getattr(self.orig_obj, attr)
+        if callable(orig_attr):
+            wrapper = self._wrapper_cache.get(attr)
+            if wrapper is None:
+                @functools.wraps(orig_attr)
+                def _wrapper(*args, **kwargs):
+                    with self.lock:
+                        return orig_attr(*args, **kwargs)
+                self._wrapper_cache[attr] = wrapper = _wrapper
```

Comment:
Is this line a typo?

Reply:
no, `self._wrapper_cache` is the cache, `wrapper` is the return value, and `_wrapper` is the local variable.
it's short for
```
self._wrapper_cache[attr] = _wrapper
wrapper = _wrapper
```